File: frontend/src/lib/package.ts
/**
 * Preview local do pacote de materiais — espelha as mesmas regras de
 * `backend/app/services/package.py::compose_package` (placeholders, filtro
 * de idioma, formatação de linha, limite de 10 links) para o operador ver o
 * texto antes de salvar, sem round-trip de rede.
 *
 * Diferenças propositais em relação ao backend (o servidor sempre revalida
 * antes de enviar de verdade):
 * - Não filtra `active`/produto: os itens já chegam filtrados por produto e
 *   ativos via GET /api/materials (lib/materials.ts).
 * - Ainda assim filtra por idioma e por url presente, para o preview bater
 *   com o que o servidor vai realmente montar.
 */
import { EmailLanguage } from "./types";
import { MaterialItem } from "./materials";

export const MAX_MATERIAL_LINKS = 10;

const LANGUAGE_TO_MATERIAL_LANG: Record<EmailLanguage, string> = {
  "pt-BR": "PT",
  en: "ENG",
  es: "ESP",
};

const PLACEHOLDER_RE = /\{(nome|primeiro_nome|evento|produto)\}/g;

export interface PackagePreviewInput {
  contactName?: string | null;
  eventTag?: string | null;
  productLabel: string;
  language: EmailLanguage;
  materialIds: number[];
  /** Catálogo candidato — itens do produto selecionado, em todos os idiomas. */
  materials: MaterialItem[];
  templateBody: string;
}

export interface PackagePreviewResult {
  text: string;
  html: string;
  warnings: string[];
  materialIdsUsed: number[];
}

function renderPlaceholders(template: string, context: Record<string, string>): string {
  const rendered = template.replace(PLACEHOLDER_RE, (_match, key: string) => context[key] ?? "");

  const cleanedLines: string[] = [];
  for (const rawLine of rendered.split("\n")) {
    let line = rawLine.replace(/ {2,}/g, " ").trim();
    line = line.replace(/ +([.,!?;:])/g, "$1");
    if (line === "" && cleanedLines.length > 0 && cleanedLines[cleanedLines.length - 1] === "") {
      continue;
    }
    cleanedLines.push(line);
  }
  return cleanedLines.join("\n").trim();
}

function materialLabel(m: MaterialItem): string {
  if (m.kind === "evento") {
    const date = m.meta?.date;
    const location = m.meta?.location;
    const extra = [date, location].filter(Boolean).join(" — ");
    if (extra) return `${m.label} (${extra})`;
  }
  return m.label;
}

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

function selectMaterials(
  materials: MaterialItem[],
  materialIds: number[],
  materialLang: string,
): { included: MaterialItem[]; warnings: string[] } {
  const warnings: string[] = [];
  const byId = new Map(materials.map((m) => [m.id, m]));
  let included: MaterialItem[] = [];

  for (const id of materialIds) {
    const m = byId.get(id);
    if (!m) {
      warnings.push(`material ${id} não encontrado — ignorado`);
      continue;
    }
    if (!m.url) {
      warnings.push(`material ${id} (${m.label}) sem url — ignorado`);
      continue;
    }
    if (m.language !== null && m.language !== materialLang) {
      warnings.push(
        `material ${id} (${m.label}) em idioma ${m.language}, pacote em ${materialLang} — ignorado`,
      );
      continue;
    }
    included.push(m);
  }

  if (included.length > MAX_MATERIAL_LINKS) {
    warnings.push(
      `pacote tinha mais de ${MAX_MATERIAL_LINKS} materiais — cortado na ordem recebida`,
    );
    included = included.slice(0, MAX_MATERIAL_LINKS);
  }

  return { included, warnings };
}

/** Monta {text, html, warnings, materialIdsUsed} para preview local. */
export function previewPackage(input: PackagePreviewInput): PackagePreviewResult {
  const name = (input.contactName || "").trim();
  const primeiroNome = name ? name.split(" ")[0] : "";
  const evento = (input.eventTag || "").trim();

  const context: Record<string, string> = {
    nome: name,
    primeiro_nome: primeiroNome,
    evento,
    produto: input.productLabel,
  };
  const body = renderPlaceholders(input.templateBody, context);

  const materialLang = LANGUAGE_TO_MATERIAL_LANG[input.language];
  const { included, warnings } = selectMaterials(input.materials, input.materialIds, materialLang);

  const lines = included.map((m) => `${materialLabel(m)} — ${m.url}`);
  const text = [body, lines.join("\n")].filter((p) => p).join("\n\n").trim();

  const paragraphs = body.split("\n\n").filter((p) => p.trim());
  const htmlParts = paragraphs.map(
    (p) => `<p>${escapeHtml(p).replace(/\n/g, "<br>")}</p>`,
  );
  if (included.length > 0) {
    const items = included.map(
      (m) => `<li><a href="${escapeHtml(m.url || "")}">${escapeHtml(materialLabel(m))}</a></li>`,
    );
    htmlParts.push(`<ul>${items.join("")}</ul>`);
  }
  const html =
    '<!DOCTYPE html><html><body style="font-family:sans-serif;font-size:14px">' +
    htmlParts.join("") +
    "</body></html>";

  return {
    text,
    html,
    warnings,
    materialIdsUsed: included.map((m) => m.id),
  };
}

/**
 * Pré-seleção do produto do pacote a partir da classificação do contato:
 * - exatamente 1 produto marcado → usa ele;
 * - vários → o primeiro na ordem de `productOrder` (mesma ordem canônica de
 *   GET /api/materials, que já reflete PRODUCTS_ORDER do backend);
 * - nenhum → null (= legado, mídia kit fixo).
 */
export function pickDefaultProduct(
  classificacao: Record<string, string | null>,
  productOrder: string[],
): string | null {
  const active = productOrder.filter((key) => classificacao[key]);
  return active.length > 0 ? active[0] : null;
}

/**
 * Pré-seleção padrão de materiais: grupo "Institucional" + itens no idioma
 * escolhido (ou sem idioma, que servem para qualquer um). O operador pode
 * desmarcar/marcar manualmente depois — isso só define o estado inicial.
 */
export function defaultMaterialIds(
  groups: { name: string; items: MaterialItem[] }[],
  language: EmailLanguage,
): number[] {
  const materialLang = LANGUAGE_TO_MATERIAL_LANG[language];
  const ids: number[] = [];
  for (const group of groups) {
    for (const item of group.items) {
      const matchesGroup = group.name === "Institucional";
      const matchesLanguage = item.language === null || item.language === materialLang;
      if ((matchesGroup || matchesLanguage) && item.url) {
        ids.push(item.id);
      }
    }
  }
  return ids;
}

File: backend/app/services/package.py
"""Composição pura do pacote de materiais por e-mail.

``compose_package`` não faz I/O (sem DB, sem rede, sem chamadas externas):
recebe os dados já carregados (contato, template do produto, materiais
candidatos, assunto já resolvido) e devolve o corpo pronto para envio.
Reaproveitado hoje pelo pipeline de e-mail e, futuramente, por WhatsApp.

Placeholders suportados no corpo do template: {nome} {primeiro_nome} {evento}
{produto}. Um placeholder sem valor correspondente é removido de forma limpa
(sem deixar "undefined", chaves cruas ou espaço duplo).
"""

from __future__ import annotations

import html as html_lib
import re
from dataclasses import dataclass, field
from typing import Any, Protocol

# Máximo de links de materiais por e-mail; acima disso, corta na ordem
# recebida e reporta em warnings.
MAX_MATERIAL_LINKS = 10

# email_language do contato/pacote ("pt-BR"|"en"|"es") -> idioma dos
# materiais no catálogo ("PT"|"ENG"|"ESP"). Materiais com language=None
# entram em qualquer idioma.
LANGUAGE_TO_MATERIAL_LANG: dict[str, str] = {"pt-BR": "PT", "en": "ENG", "es": "ESP"}

# Regra de produto: a ausência de template ativo NUNCA impede o envio. Quando
# o produto ainda não tem um MessageTemplate cadastrado, ``compose_package``
# usa este texto genérico embutido por idioma (mesmos placeholders do
# template normal) e reporta em warnings — sem levantar exceção.
DEFAULT_TEMPLATE_BODY: dict[str, str] = {
    "pt-BR": (
        "Olá, {primeiro_nome}, foi um prazer estar com você no {evento}. "
        "Seguem os materiais:"
    ),
    "en": (
        "Hello {primeiro_nome}, it was a pleasure having you at {evento}. "
        "Here are the materials:"
    ),
    "es": (
        "Hola {primeiro_nome}, fue un placer contar con tu presencia en "
        "{evento}. Aquí tienes los materiales:"
    ),
}

_PLACEHOLDER_RE = re.compile(r"\{(nome|primeiro_nome|evento|produto)\}")
_MULTI_SPACE_RE = re.compile(r" {2,}")
_SPACE_BEFORE_PUNCT_RE = re.compile(r" +([.,!?;:])")


class MaterialLike(Protocol):
    """Formato mínimo esperado para cada material candidato (duck-typed).

    Compatível tanto com ``app.db_models.Material`` quanto com fixtures de
    teste (dataclass/namedtuple simples).
    """

    id: int
    label: str
    kind: str  # link | evento
    language: str | None  # PT | ENG | ESP | None
    url: str | None
    active: bool
    meta: dict[str, Any]


@dataclass
class PackageResult:
    subject: str
    text: str
    html: str
    warnings: list[str] = field(default_factory=list)
    material_ids_used: list[int] = field(default_factory=list)


def _render_placeholders(template: str, context: dict[str, str]) -> str:
    """Substitui os placeholders suportados; limpa o que sobrar de um valor vazio."""

    def _sub(match: re.Match[str]) -> str:
        return context.get(match.group(1), "") or ""

    rendered = _PLACEHOLDER_RE.sub(_sub, template)

    cleaned_lines: list[str] = []
    for raw_line in rendered.split("\n"):
        line = _MULTI_SPACE_RE.sub(" ", raw_line).strip()
        line = _SPACE_BEFORE_PUNCT_RE.sub(r"\1", line)
        # Evita múltiplas linhas em branco seguidas (placeholder era o
        # conteúdo inteiro de uma linha do template).
        if line == "" and cleaned_lines and cleaned_lines[-1] == "":
            continue
        cleaned_lines.append(line)

    return "\n".join(cleaned_lines).strip()


def _select_materials(
    materials: list[MaterialLike],
    product_key: str,
    material_ids: list[int],
    material_lang: str | None,
) -> tuple[list[MaterialLike], list[str]]:
    """Filtra materiais selecionados: ativo + url + produto (se souber) + idioma.

    Preserva a ordem de ``material_ids``. Retorna (incluídos, warnings) —
    ids inválidos/inativos/sem url/idioma incompatível são reportados em
    warnings e excluídos, nunca silenciosamente incluídos.
    """
    warnings: list[str] = []
    by_id = {m.id: m for m in materials}
    included: list[MaterialLike] = []

    for mid in material_ids:
        m = by_id.get(mid)
        if m is None:
            warnings.append(f"material {mid} não encontrado — ignorado")
            continue
        m_product_key = getattr(m, "product_key", product_key)
        if m_product_key != product_key:
            warnings.append(
                f"material {mid} não pertence ao produto '{product_key}' — ignorado"
            )
            continue
        if not m.active:
            warnings.append(f"material {mid} ({m.label}) está inativo — ignorado")
            continue
        if not m.url:
            warnings.append(f"material {mid} ({m.label}) sem url — ignorado")
            continue
        if (
            m.language is not None
            and material_lang is not None
            and m.language != material_lang
        ):
            warnings.append(
                f"material {mid} ({m.label}) em idioma {m.language}, "
                f"pacote em {material_lang} — ignorado"
            )
            continue
        included.append(m)

    if len(included) > MAX_MATERIAL_LINKS:
        warnings.append(
            f"pacote tinha mais de {MAX_MATERIAL_LINKS} materiais — "
            f"cortado na ordem recebida"
        )
        included = included[:MAX_MATERIAL_LINKS]

    return included, warnings


def _material_label(m: MaterialLike) -> str:
    """Label de exibição; itens 'evento' ganham data/local quando houver."""
    if m.kind == "evento":
        meta = m.meta or {}
        date = meta.get("date")
        location = meta.get("location")
        extra = " — ".join(x for x in (date, location) if x)
        if extra:
            return f"{m.label} ({extra})"
    return m.label


def _build_text(body: str, materials: list[MaterialLike]) -> str:
    parts = [body] if body else []
    if materials:
        lines = [f"{_material_label(m)} — {m.url}" for m in materials]
        parts.append("\n".join(lines))
    return "\n\n".join(p for p in parts if p).strip()


def _build_html(body: str, materials: list[MaterialLike]) -> str:
    paragraphs = [p for p in body.split("\n\n") if p.strip()]
    parts = [
        f"<p>{html_lib.escape(p).replace(chr(10), '<br>')}</p>" for p in paragraphs
    ]
    if materials:
        items = []
        for m in materials:
            label = html_lib.escape(_material_label(m))
            url = html_lib.escape(m.url or "", quote=True)
            items.append(f'<li><a href="{url}">{label}</a></li>')
        parts.append("<ul>" + "".join(items) + "</ul>")
    return (
        '<!DOCTYPE html><html><body style="font-family:sans-serif;font-size:14px">'
        + "".join(parts)
        + "</body></html>"
    )


def compose_package(
    contact: Any,
    product_key: str,
    product_label: str,
    language: str,
    material_ids: list[int],
    materials: list[MaterialLike],
    template_body: str | None,
    subject: str,
) -> PackageResult:
    """Monta {subject, text, html, warnings} do e-mail de pacote.

    Args:
        contact: objeto com atributos ``name``/``event_tag`` (ScannedContact
            ou qualquer duck-type equivalente).
        product_key: chave do produto selecionado (ex: "cimi_360").
        product_label: label de exibição do produto, usado no placeholder
            {produto} (ex: "CIMI 360").
        language: idioma do envio ("pt-BR"|"en"|"es").
        material_ids: ids de materiais marcados pelo operador, na ordem de
            exibição escolhida.
        materials: catálogo candidato (materiais do produto já carregados do
            DB pelo chamador — esta função não faz I/O).
        template_body: corpo do MessageTemplate do produto, com placeholders.
            ``None``/vazio = produto sem template ativo — usa o texto
            genérico embutido (``DEFAULT_TEMPLATE_BODY``) para o idioma do
            envio e reporta em warnings. Nunca levanta exceção.
        subject: assunto já resolvido pelo chamador (reaproveita
            SUBJECTS[idioma] do pipeline legado).
    """
    name = (getattr(contact, "name", None) or "").strip()
    primeiro_nome = name.split()[0] if name else ""
    evento = (getattr(contact, "event_tag", None) or "").strip()

    context = {
        "nome": name,
        "primeiro_nome": primeiro_nome,
        "evento": evento,
        "produto": product_label,
    }

    warnings: list[str] = []
    if template_body and template_body.strip():
        body_source = template_body
    else:
        body_source = DEFAULT_TEMPLATE_BODY.get(
            language, DEFAULT_TEMPLATE_BODY["pt-BR"]
        )
        warnings.append("template padrão usado")

    body = _render_placeholders(body_source, context)

    material_lang = LANGUAGE_TO_MATERIAL_LANG.get(language)
    included, material_warnings = _select_materials(
        materials, product_key, material_ids, material_lang
    )
    warnings.extend(material_warnings)

    return PackageResult(
        subject=subject,
        text=_build_text(body, included),
        html=_build_html(body, included),
        warnings=warnings,
        material_ids_used=[m.id for m in included],
    )
